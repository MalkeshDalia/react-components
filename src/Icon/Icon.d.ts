import { CSSProperties, ComponentClass, MouseEventHandler } from 'react';
import { Type } from 'tcomb';

export type IconProps = {
  icon?: string,
  color?: string,
  paths?: number, // it was PositiveInteger
  onClick?: MouseEventHandler<HTMLElement>,
  className?: string,
  style?: CSSProperties,
};

export const Props: {
  [key: string]: Type<any>
};

declare const Icon: ComponentClass<IconProps>;
export default Icon;
