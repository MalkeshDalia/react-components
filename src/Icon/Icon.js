import React from 'react';
import range from 'lodash/range';
import { skinnable, props, t } from '../utils';
import cx from 'classnames';

const PositiveInteger = t.refinement(t.Number, x => x % 1 === 0 && x > 0, 'PositiveInteger');

export const Props = {
  icon: t.maybe(t.Str),
  color: t.maybe(t.String),
  className: t.maybe(t.Str),
  style: t.maybe(t.Obj),
  paths: t.maybe(PositiveInteger),
  onClick: t.maybe(t.Func)
};

/** An icon
 * @param icon - icon className
 * @param color - icon color
 * @param paths - number of paths the icon is composed of
 * @param onClick - onClick callback
 */
@skinnable()
@props(Props)
export default class Icon extends React.PureComponent {

  static defaultProps = {
    paths: 1,
    onClick: () => {}
  };

  getLocals(props) {
    const { className, icon, color, style = {} } = props;
    return {
      ...props,
      className: cx('icon', `icon-${icon}`, className),
      style: { ...style, color: color || style.color }
    };
  }

  template({ icon, className, style, onClick, paths }) {
    return icon ? (
      <i className={className} style={style} onClick={onClick}>
        {paths > 1 && range(paths).map(k => <span className={`path${k + 1}`} key={k} /> )}
      </i>
    )
    : null;
  }

}
