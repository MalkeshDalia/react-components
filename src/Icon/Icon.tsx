import * as React from 'react';
import range = require('lodash/range');
import { props, t } from '../utils';
import * as cx from 'classnames';

const PositiveInteger = t.refinement(t.Number, x => x % 1 === 0 && x > 0, 'PositiveInteger');

export const Props = {
  icon: t.maybe(t.String),
  color: t.maybe(t.String),
  className: t.maybe(t.String),
  style: t.maybe(t.Object),
  paths: t.maybe(PositiveInteger),
  onClick: t.maybe(t.Function)
};

export type IconDefaultProps = {
  /** number of paths the icon is composed of */
  paths: number,
  /** onClick callback*/
  onClick: React.MouseEventHandler<HTMLElement>,
  /** an optional style object to pass to top level element of the component */
  style: React.CSSProperties
};

export type IconRequiredProps = {
  /** icon className */
  icon?: string,
  /** icon color */
  color?: string,
  /** an optional class name to pass to top level element of the component */
  className?: string,
};

export namespace Icon {
  export type Props = IconRequiredProps & Partial<IconDefaultProps>;
}
type IconDefaultedProps = IconRequiredProps & IconDefaultProps;

/** An icon */
@props(Props)
export class Icon extends React.PureComponent<Icon.Props> {

  static defaultProps: IconDefaultProps = {
    paths: 1,
    onClick: () => {},
    style: {}
  };

  render() {
    const { paths, onClick, className: _className, icon, color, style: _style } = this.props as IconDefaultedProps;
    const className = cx('icon', `icon-${icon}`, _className);
    const style =  { ..._style, color: color || _style.color };

    return icon ? (
      <i className={className} style={style} onClick={onClick}>
        {paths > 1 && range(paths).map(k => <span className={`path${k + 1}`} style={{ fontSize: 'inherit' }} key={k} /> )}
      </i>
    ) : null;
  }

}
