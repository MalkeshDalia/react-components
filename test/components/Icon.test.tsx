import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Icon from '../../src/Icon';

describe('Icon', () => {

  it('renders correctly', () => {
    const component = renderer.create(
      <Icon icon='foo' />
    );
    expect(component).toMatchSnapshot();
  });

  it('renders null if no icon is passed', () => {
    const component = renderer.create(
      <Icon />
    );
    expect(component.toJSON()).toBeNull();
  });

  it('renders no child when path is 1', () => {
    const component = shallow(
      <Icon icon='foo' paths={1} />
    );
    expect(component.children().length).toBe(0);
  });

  it('renders a child for each path', () => {
    const component = shallow(
      <Icon icon='foo' paths={4} />
    );
    expect(component.children().length).toBe(4);
    component.children().forEach((child, i) => {
      expect(child.prop('className')).toBe(`path${i + 1}`);
    });
  });

  it('calls onClick once when clicked', () => {
    const onClick = jest.fn();
    const component = shallow(
      <Icon icon='foo' onClick={onClick} />
    );
    component.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });


  it('computes className', () => {
    const component = shallow(
      <Icon icon='foo' className='myicon' />
    );
    expect(component.hasClass('icon')).toBe(true);
    expect(component.hasClass('icon-foo')).toBe(true);
    expect(component.hasClass('myicon')).toBe(true);
  });

  it('computes color when not provided', () => {
    const component = shallow(
      <Icon icon='foo' />
    );
    expect(component.prop('style').color).toBeUndefined();
  });

  it('computes color when provided as prop', () => {
    const component = shallow(
      <Icon icon='foo' color='blue' />
    );
    expect(component.prop('style').color).toBe('blue');
  });

  it('computes color when provided as style', () => {
    const component = shallow(
      <Icon icon='foo' style={{ color: 'blue' }} />
    );
    expect(component.prop('style').color).toBe('blue');
  });

  it('overrides color form style if provided as prop', () => {
    const component = shallow(
      <Icon icon='foo' color='red' style={{ color: 'blue' }} />
    );
    expect(component.prop('style').color).toBe('red');
  });

});
