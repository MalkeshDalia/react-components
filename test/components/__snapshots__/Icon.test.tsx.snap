// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly 1`] = `
<i
  className="icon icon-foo"
  onClick={[Function]}
  style={
    Object {
      "color": undefined,
    }
  }
/>
`;
